tests/test_cleanup.py

```diff
@@ -487,3 +507,810 @@ def test_file_vdi_delete(self, mock_lock):
+                             lines=[2.0])
+
+        # File exists 3 values
+        self.getStorageSpeed(mock_isFile, sr, fakeFile, True, 3.0, 1,
+                             lines=[1.0, 2.0, 6.0])
+
+    def speedFileSetup(self, sr, FakeFile, mock_isFile, isFile):
+        expectedPath = cleanup.SPEED_LOG_ROOT.format(uuid=sr.uuid)
+        mock_isFile.return_value = isFile
+        FakeFile.writelines.reset_mock()
+        FakeFile.write.reset_mock()
+        FakeFile.readlines.reset_mock()
+        FakeFile.close.reset_mock()
+        FakeFile.seek.reset_mock()
+        return expectedPath
+
+    def writeSpeedFile(self, sr, speed, mock_isFile, isFile, mock_open,
+                       mock_chmod, write=None, writeLines=None, readLines=None,
+                       openOp="r+"):
+        mock_open.reset_mock()
+        mock_chmod.reset_mock()
+        expectedPath = self.speedFileSetup(sr, FakeFile, mock_isFile, isFile)
+        FakeFile.readlines.return_value = readLines
+        sr.writeSpeedToFile(speed)
+        mock_open.assert_called_with(expectedPath, openOp)
+        if openOp == "w":
+            mock_chmod.assert_called_with(expectedPath, stat.S_IRWXU)
+        if write:
+            FakeFile.write.assert_called_with(write)
+        if writeLines:
+            FakeFile.seek.assert_called_with(0)
+            FakeFile.writelines.assert_called_with(writeLines)
+        self.assertEqual(FakeFile.close.call_count, 1)
+
+    @mock.patch("__builtin__.open",
+                autospec=True)
+    @mock.patch("os.path.isfile", autospec=True)
+    @mock.patch("os.chmod", autospec=True)
+    def test_writeSpeedToFile(self, mock_chmod, mock_isFile, mock_open):
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        FakeFile = self.makeFakeFile()
+        mock_open.return_value = FakeFile
+
+        # File does not exist
+        self.writeSpeedFile(sr, 1.8, mock_isFile, False, mock_open, mock_chmod,
+                            write="1.8\n", openOp="w")
+
+        # File does exist but empty (Should not happen)
+        readLines = []
+        writeLines = ["1.8\n"]
+        self.writeSpeedFile(sr, 1.8, mock_isFile, True, mock_open, mock_chmod,
+                            readLines=readLines, writeLines=writeLines)
+
+        # File does exist, exception fired, make sure close fd.
+        expectedPath = self.speedFileSetup(sr, FakeFile, mock_isFile, True)
+        FakeFile.readlines.side_effect = Exception
+        with self.assertRaises(Exception):
+            sr.writeSpeedToFile(1.8)
+        mock_open.assert_called_with(expectedPath, 'r+')
+        self.assertEqual(FakeFile.close.call_count, 1)
+        FakeFile.readlines.side_effect = None
+
+        # File does exist
+        readLines = ["1.9\n", "2.1\n", "3\n"]
+        writeLines = ["1.9\n", "2.1\n", "3\n", "1.8\n"]
+        self.writeSpeedFile(sr, 1.8, mock_isFile, True, mock_open, mock_chmod,
```

Comment:
What are these tests achieving, just code coverage and no validation of functionality?

Reply:
I have refactored the common assert code into writeSpeedFile.

In other test frameworks this pattern is ok, but perhaps there is something wrong with doing this in python unittest.

Reply:
No, that's fine, I just hadn't seen that was what was happening :)